File: examples/calculator-python-layered/app/calculator/service/calculator_service.py
from dataclasses import dataclass

from app.calculator.domain.calculation import Calculation, CalculationId, Operator
from app.calculator.repository.calculation_repository import CalculationRepository


@dataclass(frozen=True)
class CalculationResult:
    calculation_id: str
    operand_a: float
    operator: str
    operand_b: float
    result: float


class CalculatorService:
    def __init__(self, repository: CalculationRepository) -> None:
        self._repository = repository

    def calculate(self, operand_a: float, operator: Operator, operand_b: float) -> CalculationResult:
        calculation = Calculation.perform(CalculationId.generate(), operand_a, operator, operand_b)
        self._repository.save(calculation)
        calculation.pop_events()
        return CalculationResult(
            calculation_id=str(calculation.id.value),
            operand_a=calculation.operand_a,
            operator=calculation.operator.value,
            operand_b=calculation.operand_b,
            result=calculation.result,
        )

File: examples/calculator-java-layered/src/main/java/com/oopforge/example/layered/calculator/service/CalculatorService.java
package com.oopforge.example.layered.calculator.service;

import com.oopforge.example.layered.calculator.domain.Calculation;
import com.oopforge.example.layered.calculator.domain.CalculationId;
import com.oopforge.example.layered.calculator.domain.Operator;
import com.oopforge.example.layered.calculator.repository.CalculationRepository;
import org.springframework.stereotype.Service;

@Service
public class CalculatorService {

    private final CalculationRepository calculationRepository;

    public CalculatorService(CalculationRepository calculationRepository) {
        this.calculationRepository = calculationRepository;
    }

    public CalculationResult calculate(CalculateCommand command) {
        CalculationId id = CalculationId.generate();
        Calculation calculation = Calculation.perform(id, command.operandA(), command.operator(), command.operandB());
        calculationRepository.save(calculation);
        calculation.popEvents();
        return new CalculationResult(
                id.value().toString(),
                calculation.operandA(),
                calculation.operator(),
                calculation.operandB(),
                calculation.result());
    }

    public record CalculateCommand(double operandA, Operator operator, double operandB) {}

    public record CalculationResult(String calculationId, double operandA, Operator operator, double operandB, double result) {}
}
